test/shared/src/main/scala/zio/test/experimental/TestAspect.scala

```diff
@@ -0,0 +1,378 @@
+   */
+  def before[R, E](before: ZIO[R, E, Any]): TestAspect.Aux[
+    Nothing,
+    E,
+    Nothing,
+    E,
+    ({ type Apply[+R1] = R with R1 })#Apply,
+    ({ type Apply[+E1 <: E] = E })#Apply
+  ] =
+    around(before)(_ => ZIO.unit)
+
+  /**
+   * An aspect that runs each test on a separate fiber and prints a fiber dump
+   * if the test fails or has not terminated within the specified duration.
+   */
+  def diagnose(duration: Duration): TestAspectAtLeastR[Live with Annotations] =
+    new TestAspect[Nothing, Any, Nothing, Any] {
+      type ModifyEnv[+R] = R with Live with Annotations
+      type ModifyErr[+E] = E
+
+      def some[R, E](predicate: String => Boolean, spec: ZSpec[R, E]): ZSpec[R with Live with Annotations, E] = {
+        def diagnose[R <: Live with Annotations, E](
+          label: String,
+          test: ZIO[R, TestFailure[E], TestSuccess]
+        ): ZIO[R, TestFailure[E], TestSuccess] =
+          test.fork.flatMap { fiber =>
+            fiber.join.raceWith[R, TestFailure[E], TestFailure[E], Unit, TestSuccess](Live.live(ZIO.sleep(duration)))(
+              (exit, sleepFiber) => dump(label).when(!exit.succeeded) *> sleepFiber.interrupt *> ZIO.done(exit),
+              (_, _) => dump(label) *> fiber.join
+            )
+          }
+        def dump[E, A](label: String): URIO[Live with Annotations, Unit] =
+          Annotations.supervisedFibers.flatMap(fibers => Live.live(Fiber.putDumpStr(label, fibers.toSeq: _*)))
+        spec.transform[R with Live with Annotations, TestFailure[E], TestSuccess] {
+          case c @ Spec.SuiteCase(_, _, _) => c
+          case Spec.TestCase(label, test, annotations) =>
+            Spec.TestCase(label, if (predicate(label)) diagnose(label, test) else test, annotations)
+        }
+      }
+    }
+
+  /**
+   * An aspect that runs each test with the `TestConsole` instance in the
+   * environment set to debug mode so that console output is rendered to
+   * standard output in addition to being written to the output buffer.
+   */
+  val debug: TestAspectAtLeastR[TestConsole] =
+    new TestAspect[Nothing, Any, Nothing, Any] {
+      type ModifyEnv[+R] = R with TestConsole
+      type ModifyErr[+E] = E
+
+      def some[R, E](predicate: String => Boolean, spec: ZSpec[R, E]): ZSpec[R with TestConsole, E] =
+        spec.transform[R with TestConsole, TestFailure[E], TestSuccess] {
+          case c @ Spec.SuiteCase(_, _, _) => c
+          case Spec.TestCase(label, test, annotations) =>
+            val debug: ZIO[R with TestConsole, TestFailure[E], TestSuccess] =
+              TestConsole.debug(test)
+            Spec.TestCase(label, if (predicate(label)) debug else test, annotations)
+        }
+    }
+
+  /**
+   * An aspect that applies the specified aspect on Dotty.
+   */
+  def dotty[LowerR, UpperR, LowerE, UpperE](
+    that: TestAspect[LowerR, UpperR, LowerE, UpperE]
+  ): TestAspect[LowerR, UpperR, LowerE, UpperE] =
+    if (TestVersion.isDotty) that else ???
```

Comment:
We can't implement test aspects like this in the same way because if the test aspect provides the environment or handles errors then we can't return the original spec unchanged because it doesn't satisfy the constraints of the target type. We may need to have a less polymorphic subtype of aspects that only add environment requirements or failure modes to work with operators like this.